jwst/undersampling_correction/undersampling_correction.py

```diff
@@ -61,44 +66,79 @@ def flag_pixels(data, gdq, signal_threshold):
 
     Returns
     -------
-    gdq : int, 4D array
+    new_gdq : int, 4D array
         updated group dq array
     """
     n_ints, n_grps, n_rows, n_cols = gdq.shape
-    num_pix = n_cols * n_rows
-
-    lowest_exc_1d = np.zeros(num_pix) + n_grps
-
-    for ii_int in range(n_ints):
-        for ii_grp in range(n_grps):
-            data_1d = data[ii_int, ii_grp, :, :].reshape(num_pix)  # vectorize slice
-            gdq_1d = gdq[ii_int, ii_grp, :, :].reshape(num_pix)
-
-            wh_not_dnu = np.logical_not(gdq_1d & dqflags.group['DO_NOT_USE'])
-
-            # In the current group for all ramps, locate pixels that :
-            #  a) exceed the signal_threshold, and
-            #  b) have not been previously flagged as an exceedance, and
-            #  c) were not flagged in an earlier step as DO_NOT_USE
-            wh_exc_1d = np.where((data_1d > signal_threshold) &
-                                 (lowest_exc_1d == n_grps) & wh_not_dnu)
-
-            # ... and mark those pixels, as current group is their first exceedance
-            if len(wh_exc_1d[0] > 0):  # For ramps previously unflagged ...
-                lowest_exc_1d[wh_exc_1d] = ii_grp
-
-    # Flag current and subsequent groups
-    lowest_exc_2d = lowest_exc_1d.reshape((n_rows, n_cols))
-    for ii_int in range(n_ints):
-        for ii_grp in range(n_grps):
-            wh_set_flag = np.where(lowest_exc_2d == ii_grp)
-
-            # set arrays of components
-            yy = wh_set_flag[0]
-            xx = wh_set_flag[1]
-
-            gdq[ii_int, ii_grp:, yy, xx] = \
-                np.bitwise_or(gdq[ii_int, ii_grp:, yy, xx], dqflags.group['UNDERSAMP']
-                              | dqflags.group['DO_NOT_USE'])
-
-    return gdq
+    ncols = data.shape[3]
+    nrows = data.shape[2]
+
+    new_gdq = gdq.copy()   # Updated gdq
+
+    # Flag all exceedances with UNDERSAMP and NO_NOT_USE
+    undersamp_pix = (data > signal_threshold) == (gdq != DNU)
+    new_gdq[undersamp_pix] = np.bitwise_or(new_gdq[undersamp_pix], UNSA)
+    new_gdq[undersamp_pix] = np.bitwise_or(new_gdq[undersamp_pix], DNU)
```

Comment:
Could these 2 operations be combined in one statement by using `UNSA | DNU` as a combined last argument to `bitwise_or` (as is done below in some other places)?

Reply:
Yes, done